core/endorser/endorser.go

```diff
@@ -500,3 +507,48 @@ func CreateCCEventBytes(ccevent *pb.ChaincodeEvent) ([]byte, error) {
 func decorateLogger(logger *flogging.FabricLogger, txParams *ccprovider.TransactionParams) *flogging.FabricLogger {
 	return logger.With("channel", txParams.ChannelID, "txID", shorttxid(txParams.TxID))
 }
+
+// Using the simulation results, build the ChaincodeInterest structure that the client can pass to the discovery service
+// to get the correct endorsement policy for the chaincode(s) and any collections encountered.
+func buildChaincodeInterest(simResult *ledger.TxSimulationResults) (*pb.ChaincodeInterest, error) {
+	ccInterest := &pb.ChaincodeInterest{}
+
+	chaincodes := map[string]struct{}{}
+	if simResult.PvtSimulationResults != nil {
+		// prototype - list collections that are read from (in order to populate NoPublicReads in ChaincodeInterest)
+		for _, prws := range simResult.PvtSimulationResults.GetNsPvtRwset() {
+			for _, cprws := range prws.GetCollectionPvtRwset() {
+				ccInterest.Chaincodes = append(ccInterest.Chaincodes, &pb.ChaincodeCall{
+					Name:            prws.GetNamespace(),
+					CollectionNames: []string{cprws.GetCollectionName()},
+					NoPrivateReads:  !simResult.PrivateReads.Read(prws.Namespace, cprws.CollectionName),
+				})
+				chaincodes[prws.GetNamespace()] = struct{}{}
+			}
+		}
+	}
+	// if there are any chaincodes in the public RWset that are not in the private set, add them to the interest
+	if simResult.PubSimulationResults != nil {
+		for _, nsrws := range simResult.PubSimulationResults.GetNsRwset() {
+			if nsrws.GetNamespace() == "_lifecycle" {
+				continue
+			}
+			if _, ok := chaincodes[nsrws.GetNamespace()]; !ok {
+				ccInterest.Chaincodes = append(ccInterest.Chaincodes, &pb.ChaincodeCall{
+					Name: nsrws.GetNamespace(),
+				})
+				chaincodes[nsrws.GetNamespace()] = struct{}{}
+			}
+		}
+	}
+	if len(ccInterest.Chaincodes) > 0 {
+		for _, policyBytes := range simResult.KeySignaturePolicies {
+			policy, err := protoutil.UnmarshalSignaturePolicy(policyBytes)
+			if err != nil {
+				return nil, err
+			}
+			ccInterest.Chaincodes[0].KeyPolicies = append(ccInterest.Chaincodes[0].KeyPolicies, policy)
+		}
```

Comment:
Perhaps it won't make any difference functionally but just to confirm that you deliberately mention all the SBE in the first entry for the chaincode?

Reply:
Yes, in reality it makes no difference where they are